scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,862 @@
+            logWithContext(taskid,
+                           channel,
+                           "Run waitUntil" + ON_KEY + QUOTE + key + QUOTE + RESULT_IS + answer,
+                           Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+        return answer;
+    }
+
+    private void logWithContext(TaskId taskId, String channel, String message, Level level)
+            throws InvalidChannelException {
+        String channelHeader = "[Synchronization]" + (channel != null ? "[Channel=" + channel + "] " : "");
+        switch (level.toInt()) {
+            case Level.TRACE_INT:
+                if (logger.isTraceEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.DEBUG_INT:
+                if (logger.isDebugEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.INFO_INT:
+                if (logger.isInfoEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.WARN_INT:
+                TaskLogger.getInstance().warn(taskId, channelHeader + message);
+                break;
+            case Level.ERROR_INT:
+                TaskLogger.getInstance().error(taskId, channelHeader + message);
+                break;
+            default:
+                // no action
+        }
+        if (logger.isTraceEnabled() && channel != null) {
+            TaskLogger.getInstance().info(taskId, channelHeader + "New channel content : " + getChannel(channel));
+        }
+    }
+
+    @Override
+    public void runActivity(Body body) {
+        Service service = new Service(body);
+        while (body.isActive()) {
+            try {
+                Request request = service.blockingRemoveOldest();
+                if (request == null) {
+                    return;
+                }
+                if (request.getMethodName().startsWith("wait") && !testWaitFunction(service, request)) {
+                    waitUntilQueue.add(request);
+                } else {
+                    service.serve(request);
+                    unblockWaitMethods(service);
+                }
+
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    /**
+     * Unblock all pending wait methods if predicates are met
+     * @param service active object service
+     */
+    private void unblockWaitMethods(Service service) {
+        for (Iterator<Request> iterator = waitUntilQueue.iterator(); iterator.hasNext();) {
```

Comment:
Why do you use the iterator in a loop? For clarity I think it would be better to 
- either `while` loop with the iterator
- or `for` loop on the length of the queue

Reply:
I use an iterator because I remove from the queue inside the loop. Other methods will trigger ConcurrentModificationException
"The iterators returned by this class's iterator method are fail-fast: If the deque is modified at any time after the iterator is created, in any way except through the iterator's own remove method, the iterator will generally throw a ConcurrentModificationException."